### PR TITLE
Add tether and port layer support for exec

### PR DESCRIPTION
**Pull request by hickeng**
This updates tether to check for whether a given SessionConfig should be active,
and to allow for two tiers of session - persistent & non-persistent which maps to
those created with the container and those exec'd while the container was live.
Additionally merges in @fdawg4l reload trigger as that was the easiest way to do
integration.

It also provides portlayer side package (task) that provides the Join, Bind, Unbind,
and Remove mechanics. If a task is Joined to a container while it's powered on the
resulting session is non-persistent.

This pulls in personality side changes for basic `docker exec` create/start support
from @caglar10ur but does not support the following exec mechanisms:
1. inspect
2. attach
3. start does not block until we've successful report of process launch
4. expanded tests

Fixes: #4110, #4109

### CODE REVIEW

**Comment by caglar10ur** [comment]
@hickeng any reason for not to merge this? Looks like tests are passing now as well

File: cmd/tether/main_linux.go [hunk 1]
```diff
@@ -33,22 +33,17 @@ import (
...
 
-	go func() {
-		for _ = range sigs {
-			log.Infof("Got A HUP signal! Reloading tether...")
-			tthr.Reload()
-		}
-	}()
+	// init and start the signal handler
+	if err := startSignalHandler(); err != nil {
+		log.Error(err)
```
Comment (fdawg4l) [inline cmd/tether/main_linux.go]: If you can't install the signal handle, should we `Panic()` here?

Comment (fdawg4l) [inline cmd/tether/main_linux.go]: Ah.  This func always returns nil.   Nevermind :-).

File: lib/apiservers/engine/backends/cache/container_cache.go [hunk 2]
```diff
@@ -40,6 +42,8 @@ func init() {
 		idIndex:          truncindex.NewTruncIndex([]string{}),
 		containersByID:   make(map[string]*container.VicContainer),
 		containersByName: make(map[string]*container.VicContainer),
+
```
Comment (jzt) [inline lib/apiservers/engine/backends/cache/container_cache.go]: ✂️  whitespace ✂️

File: cmd/tether/main_linux.go [hunk 3]
```diff
@@ -33,22 +33,17 @@ import (
...
 	log.SetLevel(log.DebugLevel)
 
-	go func() {
-		for _ = range sigs {
-			log.Infof("Got A HUP signal! Reloading tether...")
-			tthr.Reload()
-		}
-	}()
+	// init and start the signal handler
+	if err := startSignalHandler(); err != nil {
```
Comment (vburenin) [inline cmd/tether/main_linux.go]: why would you even want to check for error if function always return nil? I would just update that function to return nothing, so less lines of code.

File: lib/portlayer/task/common.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	task := taskS
+	if handle.Runtime != nil && handle.Runtime.PowerState != types.VirtualMachinePowerStatePoweredOff {
+		op.Debugf("Task bind configuration applies to ephemeral set")
+		task = taskE
+
+		// TODO: add check for container version - if the tether doesn't support reload/exec then
+		// this should fail
+	}
+
+	// if no task has been bound to the
```
Comment (jzt) [inline lib/portlayer/task/common.go]: This comment seems incomplete

File: lib/portlayer/task/remove.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	tasks := stasks
+	if handle.Runtime != nil && handle.Runtime.PowerState != types.VirtualMachinePowerStatePoweredOff {
+		op.Debugf("Task configuration applies to ephemeral set")
+		tasks = etasks
+
+		// TODO: add check for container version - if the tether doesn't support reload/exec then
+		// this should fail
+	}
+
+	// if no task has been bound to the
```
Comment (jzt) [inline lib/portlayer/task/remove.go]: Incomplete comment

**Comment by caglar10ur** [comment]
Where is the @sgairo - the copyright bot :) ?

**Comment by sgairo** [comment]
@caglar10ur copyright bot glitched, but is back online now

**Comment by emlin** [comment]
Lgtm

**Comment by emlin** [comment]
lgtm

**Pull request merged**